# SNPDEF

## General Considerations

This pipeline is thought for outbreak inclusion and exclusion analyses based on single nucleotide polymorphism, as well as phylogenetic inference of closely related bacterial species or serotypes. For more distantly related bacteria the amount of homoplastic events will preclude an accurate reconstruction of the evolutionary relationship of the species of interest and other methods should be considered. We also recommend to have some prior knowledge of possible mobile genetic elements, such as resistance cassettes, phages, and insertion elements of the species of interest, as they will interfere with the SNP analysis if not correctly excluded [exclusion criteria](#exclusion).

## Files required
  * Reference Genome (fasta and annotated genbank, see [Assembly and Annotation](#Assembly) workflow 1 to generate a reference from in-house sequencing or SRA files)
  * IS elements, phages, plasmids of reference genome [exclusion criteria](#exclusion)
  * Reads or assemblies from genomes of interest

## Galaxy Setup

### System Requirements
Linux or Unix based only. Some of the tools used in these workflows were developed by others and do not come in binaries for Windows.

Before installing Galaxy please check your python version (2.7 or higher). You will also need the following packages: Numpy, Pandas, Matplotlib, and Biopython. We recommend installing [anaconda python](https://docs.continuum.io/anaconda/install), which comes with many packages. 

Galaxy is an open, web-based platform for accessible, reproducible, and transparent computational biomedical research. 
Galaxy can be found on [Github](https://github.com/galaxyproject) and the manual in the [Galaxy Wiki](https://docs.galaxyproject.org/en/latest/index.html).
Galaxy can be run locally, on a server (cluster or single server), and on the cloud. If you are planning on running Galaxy on the cloud, we recommend you select a linux-based operating system.

All tools and workflows required can be retrieved from the Galaxy toolshed under the repository SNPDEF. This can be done from the browser interface directly. For more information on how to install tools, please refer to the [Galaxy Toolshed Wiki](https://docs.galaxyproject.org/en/latest/ts_api_doc.html).



## Workflows

Once installed you can upload your files through the browser or retrieve reads from NCBI based on Bioproject number with the tool "Bioproject Fastq Dump paired" into a paired read list and reference genomes with "Retrieve genome NCBI".

### Exclusion Criteria <a name="exclusion"></a>

#### Excluding Mobilome Regions

##### Excluding with a closed reference:
Use the workflow **"workflow 2a"**.
If you have a closed reference with known mobilome and repeated regions in an interval file this step can be skipped. If you don't know the mobilome use [PHASTER](http://phaster.ca/) for phage regions prediction and save regions as an interval formatted file. Use [ISFinder](https://www-is.biotoul.fr/) to predict IS elements for your species of interest. If your species carries other mobile genetic elements within the chromosome please add them to the interval file (e.g. resistance cassettes). 

##### Excluding with a draft genome:
Use the workflow **"workflow 2b"**.
For draft genomes mobile genetic elements need to be predicted like for closed references with [PHAST](http://phast.wishartlab.com/). However the predicted regions should be saved as multifasta, since intervals provided by PHAST correspond to concatenated contigs and not the actual intervals on each contig. For IS elements you can use  [ISFinder](https://www-is.biotoul.fr/). If your species carries any other mobile genetic elements please add the regions to the interval file (e.g. resistance cassettes). Draft genomes can have plasmid contigs in the assembly, therefore provide a multifasta of plasmids commonly found in the species of interest to exclude them. 

###### Example interval file
    NC_011353.1	273971	274038
    NC_011353.1	275213	276349
    NC_011353.1	302573	314525
    
### SNPDEF Reads (workflow 3)
You will need a fasta and genbank file of the reference genome, fastq files for each query genome, and an exclusion dataset generated with workflows 2a or 2b.

#### Reads-based Discovery 
The reads should be provided as a paired list (autmatically generated by fastq_dump_Bioproject_paired). SNPs will be predicted by Bowtie2 mapping & Freebayes variant calling. If you are using MiSeq reads you might have to lower the coverage to 10 in the Freebayes options.
![coverage](https://github.com/brigidar/galaxy_workflows/blob/master/coverage_fb.png)

#### SNP filtering and annotating
SNPs are compared to excluded regions for filtering. Remaining SNPs are combined into a single table and no hits and positions with ambigous nucleotides are removed (optional). The output is a filtered table and a multifasta with the curated SNPs for each query genome.

### SNPDEF Assembly (workflow 4)
You will need a fasta and genbank file of the reference genome, fasta files for each query genome, and an exclusion dataset generated with workflows 2a or 2b.

#### Contig-based Discovery 
SNP discovery is based on NUCmer with delta-filter and show-snps.

#### SNP Exclusion and Filtering contigs  
The workflow is by default threaded. Number of threads will dependon on number of SNPs (nb of threads = 30% of predicted SNPs or max amount of available threads). To change to a non-threaded version open the workflow and modify the SNP verify block.  
![change to threaded](https://github.com/brigidar/galaxy_workflows/blob/master/threaded.png)

The output is a filtered table and a multifasta with the curated SNPs for each query genome, as well as a count of SNPs per genome. 



##### Example filtered SNP table
    molecule	    refpos	 syn?	 refbase	 qbase:abht	 gene name	          gene start	 gene end	 gene length	 snps per gene	 pos in gene/	
    NZ CM000662	    101797	 NSYN	 A	         T	         ESCCO14588_RS26480	  101814	     101644	     171	         1	             18//	
    
    //ref codon  ref aa      query codon	query aa	 transition/transversion	//		maxlen:abht	 blengths:abht	 product
      TTA        L           TTT	        F	         transversion	           // 	    40	         40	             hypothetical

The table not only provides the SNP, but also the corresponding annotation from the reference genome. For contig based discovery it provides the length of the blastn hit(maxlen), as well if there are additional hits of lower quality (blenghts). Positions with multiple high quality blastn hits are excluded, as well as indels, no hits, and invariant sites.

### Assembly and Annotation (workflow 1) <a name="Assembly"></a>


**Assembly**

The workflow uses SPAdes v.3.7.1 for *de novo* assembly of reads. 
  * Reads should be imported as fastqsanger format and grouped into a paired dataset list.
  * If you are assembling longer reads please change the k-mer size in the SPADES option
     * Reads 250 bp or more use k-mer 21,33,55,77,99,127
     ![kmer option](https://github.com/brigidar/galaxy_workflows/blob/master/kmer.png)

  * Scaffolds are filtered according to size and coverage
     * Coverage or length settings should be changed according to sequencing technology (MiSeq, NextSeq, HiSeq)  
     ![coverage option](https://github.com/brigidar/galaxy_workflows/blob/master/coverage.png)

  * For more information on SPAdes please refer to the [manual] (http://spades.bioinf.spbau.ru/release3.7.0/manual.html)
  * An assembled genome is required as a reference for SNP calling.
   
**A.3 Annotation**

Annotation provided by PROKKA
  * For more information on PROKKA please refer to the [github repository](https://github.com/tseemann/prokka)
  * An annotated genbank file is required for the reference sequence only.

  
#### Phylogeny (workflow 5 optional)
SNP fasta file is required to generate a maximum likelihood phylogeny with RAxML. The ASC_GAMMA nucleotide model with 0 invariable sites and ASC correction style stamatakis is used. 

    
#### Rerooting (workflow 6 optional)
This workflow will allow you to reroot the SNP tree for better visualization on your leafs of interest.


#### Genotyping (workflow 7 optional)

The filtered table is processed to provide genotypes and summary of SNP characteristics.
![summary genotypes](https://github.com/brigidar/galaxy_workflows/blob/master/genotyper.png)

Provide the name of the reference genome in the genotyper summary tab. 
![reference genome name](https://github.com/brigidar/galaxy_workflows/blob/master/genotyper_select.png)


### Other scripts available
**Plotting of SNPs**
To plot SNPs along the genome povide the length of the genome and the bin size for the SNPs. This option is currently only available for closed genomes.
![snp location](https://github.com/brigidar/galaxy_workflows/blob/master/figure_1.png)

**Retrieve SRR reads**
This tool fetches SRR reads based on Bioproject number (NCBI) and saves it as a paired list.

**Retrieve genomes**
This tool gets genbank and fasta file for genome based on NCBI ID.

